TLS Support for Bacula

RaHoni opened the pull request:
## Add support for TLS in the Bacula module

This PR adds support to configure TLS in Bacula. 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

peterhoeg committed:
rancid: init at 3.13

GlancingMind committed:
eclipses.eclipse-dsl: Init at 2023-09

RaHoni commented:
@lovek323 and @proteansec is there anything that I should add or do I need to test something?

JulianFP commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -9,6 +9,21 @@ with lib;
+      ++ optional (tls_cfg.certificate != null) ''TLS Certificate = "${tls_cfg.certificate}";''
+      ++ [''TLS Key = "${tls_cfg.key}";'']
+      ++ optional (tls_cfg.verifyPeer != null) ''TLS Verify Peer = ${yes_no tls_cfg.verifyPeer};''
+      ++ optional (tls_cfg.allowedCN != { }) ''TLS Allowed CN = ${concatStringsSep " " (attrNames tls_cfg.allowedCN)};''
+      ++ optional (tls_cfg.caCertificateFile != null) ''TLS CA Certificate File = ${tls_cfg.caCertificateFile};''
```
```suggestion
      ++ optional (tls_cfg.allowedCN != { }) ''TLS Allowed CN = ${concatStringsSep " " tls_cfg.allowedCN};''
      ++ [''TLS CA Certificate File = "${tls_cfg.caCertificateFile}";'']
```

RaHoni replied:
As you can see at [Bacula Docs](https://www.bacula.org/11.0.x-manuals/en/main/Configuring_Director.html#SECTION0021200000000000000000) under `TLS Allowed CN = ` you are not supposed to add quotes to `CA Certificate File`.

JulianFP replied:
> As you can see at the [Bacula Docs](https://www.bacula.org/11.0.x-manuals/en/main/Configuring_Director.html#SECTION0021200000000000000000) under `TLS Allowed CN = ` you are not supposed to add quotes to `CA Certificate File`.

Correct me if I'm wrong, but while the example doesn't use quotes, `TLS CA Certificate File` accepts datatype `<Filename>` just like `TLS Key` and `TLS Certificate`. So shouldn't we treat these three options in the same way (meaning either quotes for all three of them or for non of them)?

RaHoni replied:
Yes you are Correct. But I am honestly not completely sure  witch is the better way. Since on the one hand the Documentation does not use Quotes and therefor it is the recommended style. On the other hand I am not sure about special characters like space in filenames.

RaHoni replied:
Just looked at the Documentation for the [Data Types](https://www.bacula.org/11.0.x-manuals/en/main/Customizing_Configuration_F.html#SECTION0020240000000000000000) 

> A directory is either a quoted or non-quoted string.

So lets use quoted strings just for easier handling of special characters.

JulianFP commented on nixos/modules/services/backup/bacula.nix:
```diff
@@ -108,13 +128,94 @@ let
+         '';
+      };
+      allowedCN = mkOption {
+        type = types.attrsOf types.str;
```
```suggestion
        type = types.listOf types.str;
```

JulianFP reviewed (changes requested):
Looks mostly good, however I noticed the following mistakes:

1. allowedCN should be an list of strings instead of an attrSet
2. CA Certificate File is a required option (no default defined), so it makes no sence to put it behing optional keyword
3. path of CA Certificate File was not in a string (quotes missing)

tc-kaluza committed:
maintainers: add tc-kaluza

tc-kaluza committed:
insulator2: init at 2.12.2

address comments

JulianFP reviewed (changes requested):
Besides from this one small mistake everything looks good to go. The indentation in the generated config file is all wrong, however this is irrelevant for bacula to work and was already the case before your PR. Since I also don't know an easy fix for that, I would suggest to just ignore it for now.

JulianFP reviewed (approved):
Tested it again in a VM, looks all good to me now

taranarmo committed:
passExtensions.pass-file: init at 1.0.0

RaHoni commented:
@therealr5 I see that you worked on the bacula module could you take a look at this PR and give me feedback on what to improve

Noodlez1232 committed:
maintainer: add noodlez1232

Noodlez1232 committed:
joycond-cemuhook: init at unstable-2023-08-09

Commit:
litterbox: init at 1.9

Commit:
maintainers: add ajwhouse

n00b0ss committed:
python3.pkgs.gcodepy: init at 0.1.1

n00b0ss committed:
maintainers: add n00b00s

ajs124 committed:
mpvScripts.mpris: 1.0 -> 1.1

https://github.com/hoyon/mpv-mpris/releases/tag/1.1

Commit:
badlion-client: init at 3.15.0

r-ryantm committed:
wxmaxima: 23.10.0 -> 23.11.0

Commit:
java-service-wrapper: build with latest jdk

ajs124 committed:
sope: 5.8.0 -> 5.9.0

+ fix build

ajs124 committed:
sogo: add note to also update sogo

emilytrau committed:
openocd-rp2040: override openocd package instead

r-ryantm committed:
firewalld: 2.0.1 -> 2.0.2

marsam committed:
cozette: 1.22.2 -> 1.23.1

Changelog: https://github.com/slavfox/Cozette/blob/v.1.23.1/CHANGELOG.md

marsam committed:
nodejs_21: 21.2.0 -> 21.3.0

Changelog: https://github.com/nodejs/node/releases/tag/v21.3.0

marsam committed:
millet: 0.13.5 -> 0.14.0

Diff: https://github.com/azdavis/millet/compare/v0.13.5...v0.14.0

Changelog: https://github.com/azdavis/millet/blob/v0.14.0/docs/CHANGELOG.md

philiptaron committed:
qrtool: 0.8.4 -> 0.8.5

I added `asciidoctor` to the nativeBuildInputs as I saw warnings that it
wasn't found as I built the tool.

I also stopped using `rec` syntax as the nixpkgs manual recommends
[against it](https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes).

However, `rustPlatform.buildRustPackage` doesn't allow the use of
`mkDerivation (finalAttrs: {` form, so it's just a simple `let` binding
here.

ajs124 committed:
sogo: fix build

kilianar committed:
broot: optional support for clipboard feature

This commit introduces support for the clipboard feature flag in broot,
allowing users to copy paths to the system clipboard. Enabling this
feature is done by setting withClipboard = true. However, as per the
upstream documentation, enabling this option may cause compilation
issues on certain platforms, which is why it remains disabled by
default.

kilianar committed:
broot: 1.28.1 -> 1.29.0

https://github.com/Canop/broot/releases/tag/v1.29.0

mkg20001 committed:
common-licenses: 11.1 -> 13

uku3lig committed:
vencord: 1.6.3 -> 1.6.4

NANASHI0X74 committed:
pythonPackages.qdarkstyle: 3.1 -> 3.2.3

philiptaron committed:
qrtool: add shell completions and man pages

otavio committed:
rio: 0.0.27 -> 0.0.28

Signed-off-by: Otavio Salvador <otavio@ossystems.com.br>

Kiskae committed:
libva: convert to finalAttrs

Kiskae committed:
libva: add pkg-config test

Commit:
fractal: drop drawin support

MikaelFangel committed:
changedetection-io: 0.45.7.3 -> 0.45.8.1

viraptor committed:
gemstash: 2.1.0 -> 2.7.1

philiptaron committed:
qrtool: use `rustPlatform.buildRustPackage rec {`

Rather than `let ... in` syntax per feedback from @atalii

r-ryantm committed:
python310Packages.django-import-export: 3.3.1 -> 3.3.3

pbsds committed:
certmgr: 1.6.4 -> 3.0.3, migrate to buildGoModule and by-name

spacefault committed:
osu-lazer-bin: 2023.1114.1 -> 2023.1130.0

This commit updates osu-lazer-bin from version 2023.1114.1 to
2023.1130.0. The changelogs for this update can be viewed here:
https://github.com/ppy/osu/releases/tag/2023.1130.0

acuteenvy committed:
deno: 1.38.2 -> 1.38.4

Commit:
hybridreverb2: enableParallelBuilding=true

Misaka13514 committed:
nuclei: 3.0.4 -> 3.1.0

Diff: https://github.com/projectdiscovery/nuclei/compare/v3.0.4...v3.1.0

Changelog: https://github.com/projectdiscovery/nuclei/releases/tag/v3.1.0

dotlambda committed:
corrosion: 0.4.4 -> 0.4.5

Diff: https://github.com/corrosion-rs/corrosion/compare/v0.4.4...v0.4.5

Changelog: https://github.com/corrosion-rs/corrosion/blob/v0.4.5/RELEASES.md

SuperSandro2000 committed:
onefetch: 2.18.1 -> 2.19.0

SuperSandro2000 committed:
liberasurecode: remove -Werror

r-ryantm committed:
python310Packages.drms: 0.6.4 -> 0.7.0

NickHu committed:
sway-assign-cgroups: init at 0.4.0 (#268949)

therealr5 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -801,6 +801,13 @@ Make sure to also check the many updates in the [Nixpkgs library](#sec-release-2
 - `postgresql_11` has been removed since it'll stop receiving fixes on November
   9th 2023.
 
+
```
seems unrelated. Did this land in this PR by accident? :eyes:

RaHoni replied:
Yes probably while rebasing

rouven0 reviewed (commented):
the diff looks good to me. But you probably want to redo this in 1 or 2 commits to conform to the clean commit style here.

Perhaps one that introduces the tls options and one that migrates `lib.mdDoc` to `mdDoc`

RaHoni committed:
nixos/bacula: Add support for TLS

Added the majority of TLS options used for transport encryption.

This uses a submodule since all resources share the same settings.

The documentation can be found under:
https://www.bacula.org/13.0.x-manuals/en/main/Bacula_TLS_Communications_E.html

RaHoni committed:
nixos/bacula: refactor option generation

The generation of the descriptions always used lib explicitly even though
it was already specified with 'with'.
